### Springboard Data Science - Capstone 3

#### Sentiment Analysis - Data Wrangling

##### Introduction

###### What topic does it cover?

Acknowledgements: 
The data was scraped from Booking.com. All data in the file is publicly available to everyone already. Please be noted that data is originally owned by Booking.com.

Data Context:
This dataset contains 515,000 customer reviews and scoring of 1493 luxury hotels across Europe.

The csv file contains 17 fields. The description of each field is as below:

- Hotel_Address: Address of hotel.
- Review_Date: Date when reviewer posted the corresponding review.
- Average_Score: Average Score of the hotel, calculated based on the latest comment in the last year.
- Hotel_Name: Name of Hotel
- Reviewer_Nationality: Nationality of Reviewer
- Negative_Review: Negative Review the reviewer gave to the hotel. If the reviewer does not give the negative review, then it should be: 'No Negative'
- ReviewTotalNegativeWordCounts: Total number of words in the negative review.
- Positive_Review: Positive Review the reviewer gave to the hotel. If the reviewer does not give the negative review, then it should be: 'No Positive'
- ReviewTotalPositiveWordCounts: Total number of words in the positive review.
- Reviewer_Score: Score the reviewer has given to the hotel, based on his/her experience
- TotalNumberofReviewsReviewerHasGiven: Number of Reviews the reviewers has given in the past.
- TotalNumberof_Reviews: Total number of valid reviews the hotel has.
- Tags: Tags reviewer gave the hotel.
- dayssincereview: Duration between the review date and scrape date.
- AdditionalNumberof_Scoring: There are also some guests who just made a scoring on the service rather than a review. This number indicates how many valid scores without review in there.
- lat: Latitude of the hotel
- lng: longtitude of the hotel

#### Importing Data

In [99]:
import pandas as pd 
import numpy as np

path = "C:\\Users\\hanna\\OneDrive\\Desktop\\All Folders\\Data Science\\SpringBoard\\Capstone 3\\Hotel Reviews Europe\\Hotel_Reviews.csv"

hotels = pd.read_csv(path)

In [100]:
hotels.head(3)

Hotel_Address  \
0   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
1   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
2   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   

   Additional_Number_of_Scoring Review_Date  Average_Score   Hotel_Name  \
0                           194    8/3/2017            7.7  Hotel Arena   
1                           194    8/3/2017            7.7  Hotel Arena   
2                           194   7/31/2017            7.7  Hotel Arena   

  Reviewer_Nationality                                    Negative_Review  \
0              Russia    I am so angry that i made this post available...   
1             Ireland                                         No Negative   
2           Australia    Rooms are nice but for elderly a bit difficul...   

   Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
0                                397                     1403   
1                                  0                     1403   
2                                 42                     1403   

                                     Positive_Review  \
0   Only the park outside of the hotel was beauti...   
1   No real complaints the hotel was great great ...   
2   Location was good and staff were ok It is cut...   

   Review_Total_Positive_Word_Counts  \
0                                 11   
1                                105   
2                                 21   

   Total_Number_of_Reviews_Reviewer_Has_Given  Reviewer_Score  \
0                                           7             2.9   
1                                           7             7.5   
2                                           9             7.1   

                                                Tags days_since_review  \
0  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days   
1  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days   
2  [' Leisure trip ', ' Family with young childre...            3 days   

         lat       lng  
0  52.360576  4.915968  
1  52.360576  4.915968  
2  52.360576  4.915968

#### Data Exploration

In [101]:
hotels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 515738 entries, 0 to 515737
Data columns (total 17 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   Hotel_Address                               515738 non-null  object 
 1   Additional_Number_of_Scoring                515738 non-null  int64  
 2   Review_Date                                 515738 non-null  object 
 3   Average_Score                               515738 non-null  float64
 4   Hotel_Name                                  515738 non-null  object 
 5   Reviewer_Nationality                        515738 non-null  object 
 6   Negative_Review                             515738 non-null  object 
 7   Review_Total_Negative_Word_Counts           515738 non-null  int64  
 8   Total_Number_of_Reviews                     515738 non-null  int64  
 9   Positive_Review                             515738 non-null  object 
 

Besides the "lat" and "lng" column, all of the columns have no missing values, we will work with it more as we go along. 
Also it is worth mentioning that not all columns will be useful to us, we will drop columns along the way. At the end, our goal is to find patterns between the textual reviews provided by the customers. 

In [102]:
hotels.describe().T

count         mean  \
Additional_Number_of_Scoring                515738.0   498.081836   
Average_Score                               515738.0     8.397487   
Review_Total_Negative_Word_Counts           515738.0    18.539450   
Total_Number_of_Reviews                     515738.0  2743.743944   
Review_Total_Positive_Word_Counts           515738.0    17.776458   
Total_Number_of_Reviews_Reviewer_Has_Given  515738.0     7.166001   
Reviewer_Score                              515738.0     8.395077   
lat                                         512470.0    49.442439   
lng                                         512470.0     2.823803   

                                                    std        min  \
Additional_Number_of_Scoring                 500.538467   1.000000   
Average_Score                                  0.548048   5.200000   
Review_Total_Negative_Word_Counts             29.690831   0.000000   
Total_Number_of_Reviews                     2317.464868  43.000000   
Review_Total_Positive_Word_Counts             21.804185   0.000000   
Total_Number_of_Reviews_Reviewer_Has_Given    11.040228   1.000000   
Reviewer_Score                                 1.637856   2.500000   
lat                                            3.466325  41.328376   
lng                                            4.579425  -0.369758   

                                                    25%          50%  \
Additional_Number_of_Scoring                 169.000000   341.000000   
Average_Score                                  8.100000     8.400000   
Review_Total_Negative_Word_Counts              2.000000     9.000000   
Total_Number_of_Reviews                     1161.000000  2134.000000   
Review_Total_Positive_Word_Counts              5.000000    11.000000   
Total_Number_of_Reviews_Reviewer_Has_Given     1.000000     3.000000   
Reviewer_Score                                 7.500000     8.800000   
lat                                           48.214662    51.499981   
lng                                           -0.143372     0.010607   

                                                    75%           max  
Additional_Number_of_Scoring                 660.000000   2682.000000  
Average_Score                                  8.800000      9.800000  
Review_Total_Negative_Word_Counts             23.000000    408.000000  
Total_Number_of_Reviews                     3613.000000  16670.000000  
Review_Total_Positive_Word_Counts             22.000000    395.000000  
Total_Number_of_Reviews_Reviewer_Has_Given     8.000000    355.000000  
Reviewer_Score                                 9.600000     10.000000  
lat                                           51.516288     52.400181  
lng                                            4.834443     16.429233

"Additional_Number_of_Scoring" column is based on guests who just made a scoring on the service rather than a review. This number indicates how many valid scores without review in there. However, we see values ranging from 1.0 to 2682, as we do not have a certain threshold, say 1 being worst and 10 being best, there is nothing we can make of these numerical values.

In [103]:
hotels.describe(include='object').T

count  unique  \
Hotel_Address         515738    1493   
Review_Date           515738     731   
Hotel_Name            515738    1492   
Reviewer_Nationality  515738     227   
Negative_Review       515738  330011   
Positive_Review       515738  412601   
Tags                  515738   55242   
days_since_review     515738     731   

                                                                    top  \
Hotel_Address         163 Marsh Wall Docklands Tower Hamlets London ...   
Review_Date                                                    8/2/2017   
Hotel_Name                   Britannia International Hotel Canary Wharf   
Reviewer_Nationality                                    United Kingdom    
Negative_Review                                             No Negative   
Positive_Review                                             No Positive   
Tags                  [' Leisure trip ', ' Couple ', ' Double Room '...   
days_since_review                                                1 days   

                        freq  
Hotel_Address           4789  
Review_Date             2585  
Hotel_Name              4789  
Reviewer_Nationality  245246  
Negative_Review       127890  
Positive_Review        35946  
Tags                    5101  
days_since_review       2585

In [104]:
hotels.duplicated().sum()

526

In [105]:
hotels[hotels.duplicated()].head(5)

Hotel_Address  \
23704  100 110 Euston Road Camden London NW1 2AJ Unit...   
61876  16 22 Great Russell Street Camden London WC1B ...   
68967        167 rue de Rome 17th arr 75017 Paris France   
68968        167 rue de Rome 17th arr 75017 Paris France   
68969        167 rue de Rome 17th arr 75017 Paris France   

       Additional_Number_of_Scoring Review_Date  Average_Score  \
23704                           728   2/25/2017            8.9   
61876                           300   7/27/2017            9.0   
68967                            11   6/23/2017            6.8   
68968                            11   4/27/2017            6.8   
68969                            11    3/4/2017            6.8   

                      Hotel_Name Reviewer_Nationality  \
23704  Pullman London St Pancras               China    
61876       The Bloomsbury Hotel              Israel    
68967              Villa Eugenie         Netherlands    
68968              Villa Eugenie      United Kingdom    
68969              Villa Eugenie        South Africa    

                                         Negative_Review  \
23704                                               none   
61876                                        No Negative   
68967   Location is very old not invested in past yea...   
68968   This hotel I don t think so is 4 star B B hav...   
68969   The staff were not helpful at all It took us ...   

       Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
23704                                  2                     3168   
61876                                  0                     1254   
68967                                 41                      165   
68968                                 93                      165   
68969                                 69                      165   

                                         Positive_Review  \
23704   good location and super nice staff room is bi...   
61876   The attention received by Sebastian and his t...   
68967   Location is central accessible by train parki...   
68968                                            Nothing   
68969                                  Near bus terminal   

       Review_Total_Positive_Word_Counts  \
23704                                 13   
61876                                 12   
68967                                 12   
68968                                  2   
68969                                  4   

       Total_Number_of_Reviews_Reviewer_Has_Given  Reviewer_Score  \
23704                                          39             9.6   
61876                                           4             9.6   
68967                                          24             4.6   
68968                                          10             2.5   
68969                                           4             5.0   

                                                    Tags days_since_review  \
23704  [' Business trip ', ' Solo traveler ', ' Class...           159 day   
61876  [' Leisure trip ', ' Couple ', ' Superior Doub...            7 days   
68967  [' Business trip ', ' Solo traveler ', ' Singl...           41 days   
68968  [' Business trip ', ' Solo traveler ', ' Singl...           98 days   
68969  [' Leisure trip ', ' Couple ', ' Twin Room ', ...           152 day   

             lat       lng  
23704  51.528677 -0.128349  
61876  51.517167 -0.129053  
68967  48.887128  2.314205  
68968  48.887128  2.314205  
68969  48.887128  2.314205

There is no conclusive evidence that the rows are fully duplicated. We see that some entries are such as the address, number of scoring, but upon a full glance of the row, the entries are not fully duplicated, hence we will not be removing them.

let us convert the Review_Date column from object to datetime, and make three columns for day, month, and year. 

In [106]:
hotels['Review_Date'] = pd.to_datetime(hotels['Review_Date'])

In [107]:
hotels['Review_Date'].dtypes

dtype('<M8[ns]')

In [108]:
hotels['Month'] = pd.DatetimeIndex(hotels['Review_Date']).month
hotels['Year'] = pd.DatetimeIndex(hotels['Review_Date']).year
hotels['Day'] = pd.DatetimeIndex(hotels['Review_Date']).day

hotels.drop('Review_Date', axis = 1, inplace = True)

In [109]:
hotels.head(4)

Hotel_Address  \
0   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
1   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
2   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
3   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   

   Additional_Number_of_Scoring  Average_Score   Hotel_Name  \
0                           194            7.7  Hotel Arena   
1                           194            7.7  Hotel Arena   
2                           194            7.7  Hotel Arena   
3                           194            7.7  Hotel Arena   

  Reviewer_Nationality                                    Negative_Review  \
0              Russia    I am so angry that i made this post available...   
1             Ireland                                         No Negative   
2           Australia    Rooms are nice but for elderly a bit difficul...   
3      United Kingdom    My room was dirty and I was afraid to walk ba...   

   Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
0                                397                     1403   
1                                  0                     1403   
2                                 42                     1403   
3                                210                     1403   

                                     Positive_Review  \
0   Only the park outside of the hotel was beauti...   
1   No real complaints the hotel was great great ...   
2   Location was good and staff were ok It is cut...   
3   Great location in nice surroundings the bar a...   

   Review_Total_Positive_Word_Counts  \
0                                 11   
1                                105   
2                                 21   
3                                 26   

   Total_Number_of_Reviews_Reviewer_Has_Given  Reviewer_Score  \
0                                           7             2.9   
1                                           7             7.5   
2                                           9             7.1   
3                                           1             3.8   

                                                Tags days_since_review  \
0  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days   
1  [' Leisure trip ', ' Couple ', ' Duplex Double...            0 days   
2  [' Leisure trip ', ' Family with young childre...            3 days   
3  [' Leisure trip ', ' Solo traveler ', ' Duplex...            3 days   

         lat       lng  Month  Year  Day  
0  52.360576  4.915968      8  2017    3  
1  52.360576  4.915968      8  2017    3  
2  52.360576  4.915968      7  2017   31  
3  52.360576  4.915968      7  2017   31

In [110]:
hotels['Year'].value_counts()

2016    264403
2017    156808
2015     94527
Name: Year, dtype: int64

In [111]:
hotels['Reviewer_Nationality'].value_counts().sort_values(ascending=False).head(20)

 United Kingdom               245246
 United States of America      35437
 Australia                     21686
 Ireland                       14827
 United Arab Emirates          10235
 Saudi Arabia                   8951
 Netherlands                    8772
 Switzerland                    8678
 Germany                        7941
 Canada                         7894
 France                         7296
 Israel                         6610
 Italy                          6114
 Belgium                        6031
 Turkey                         5444
 Kuwait                         4920
 Spain                          4737
 Romania                        4552
 Russia                         3900
 South Africa                   3821
Name: Reviewer_Nationality, dtype: int64

We see that we have 3 years worth of reviews, and we learned that Review_Date has 731 unique values which corroborates with the fact that we have 3 years worth of reviews. 

In [112]:
print(len(hotels[hotels['Negative_Review'] == 'No Negative']))
print(len(hotels[hotels['Positive_Review'] == 'No Positive']))

127890
35946


In [113]:
country = hotels['Hotel_Address'].map(lambda x: x.split()[-1])
country

0         Netherlands
1         Netherlands
2         Netherlands
3         Netherlands
4         Netherlands
             ...     
515733        Austria
515734        Austria
515735        Austria
515736        Austria
515737        Austria
Name: Hotel_Address, Length: 515738, dtype: object

While looking in excel, there were some inconsistencies with United Kingdom written as Kingdom. We can look through it, and we can map the word United Kingdon whereever we find the word kingdom.

In [114]:
hotels['Hotel_Address'].map(lambda x: x.split()[-1])[500:600]

500    Kingdom
501    Kingdom
502    Kingdom
503    Kingdom
504    Kingdom
        ...   
595    Kingdom
596    Kingdom
597    Kingdom
598    Kingdom
599    Kingdom
Name: Hotel_Address, Length: 100, dtype: object

In [115]:
country = country.map(lambda x: 'United Kingdom' if x == 'Kingdom' else x)
country[500:600]

500    United Kingdom
501    United Kingdom
502    United Kingdom
503    United Kingdom
504    United Kingdom
            ...      
595    United Kingdom
596    United Kingdom
597    United Kingdom
598    United Kingdom
599    United Kingdom
Name: Hotel_Address, Length: 100, dtype: object

Let us check for the names of countries and which country has had the most reviews. 

In [116]:
country.value_counts()

United Kingdom    262301
Spain              60149
France             59928
Netherlands        57214
Austria            38939
Italy              37207
Name: Hotel_Address, dtype: int64

Using this, we can subsitute a new column country to illustrate the country of the hotel, given its address. 

In [117]:
hotels['country'] = country

In [118]:
hotels.tail(3)

Hotel_Address  \
515735  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   
515736  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   
515737  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...   

        Additional_Number_of_Scoring  Average_Score             Hotel_Name  \
515735                           168            8.1  Atlantis Hotel Vienna   
515736                           168            8.1  Atlantis Hotel Vienna   
515737                           168            8.1  Atlantis Hotel Vienna   

       Reviewer_Nationality  \
515735               Egypt    
515736              Mexico    
515737             Hungary    

                                          Negative_Review  \
515735   The ac was useless It was a hot week in vienn...   
515736                                        No Negative   
515737       I was in 3rd floor It didn t work Free Wife    

        Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
515735                                 19                     2823   
515736                                  0                     2823   
515737                                 13                     2823   

                                          Positive_Review  \
515735                                        No Positive   
515736   The rooms are enormous and really comfortable...   
515737                               staff was very kind    

        Review_Total_Positive_Word_Counts  \
515735                                  0   
515736                                 25   
515737                                  6   

        Total_Number_of_Reviews_Reviewer_Has_Given  Reviewer_Score  \
515735                                           3             2.5   
515736                                           3             8.8   
515737                                           1             8.3   

                                                     Tags days_since_review  \
515735  [' Leisure trip ', ' Family with older childre...           715 day   
515736  [' Leisure trip ', ' Group ', ' Standard Tripl...           717 day   
515737  [' Leisure trip ', ' Family with young childre...           725 day   

              lat        lng  Month  Year  Day  country  
515735  48.203745  16.335677      8  2015   19  Austria  
515736  48.203745  16.335677      8  2015   17  Austria  
515737  48.203745  16.335677      8  2015    9  Austria

The Tags column seems to have a pattern, the first element of the list tells us whether it was a business trip or Leisure trip, second is for the amount of people, third is for the type of room, fourth is the number of nights stayed, and fifth is whether the tags were submitted via a phone device. We can make a copy so we can see if the list is polluted of values interchanging their position hence making it harder to extract information. If that is the case, we will extract the values using exact words, such as 'Leisure trip', 'Group', 'Family with young/old kids, etc. 

In [119]:

hotels_copy = hotels.copy()

tags = hotels_copy['Tags'].str.split(',', expand = True)
# print("Printing Tags after splitting")
# print(tags)


# Providing a structure to see if values change positions or not. 
hotels_copy[['Trip_Type', 'People', 'Room', 'Duration', 'Device', 'Other']] = hotels_copy['Tags'].str.split(',', expand=True)

print("")
print('Printing Columns')
print(hotels_copy.columns)

#Dropping unnecessary columns, and the Tags column itself. 
hotels_copy.drop(['Tags', 'Device', 'Other'], axis = 1, inplace = True)

print("")
print("Printing After Dropping the Columns")
print(hotels_copy.columns)

#Stripping the special characters
hotels_copy['Trip_Type'] = hotels_copy['Trip_Type'].str.replace('[', '')
hotels_copy['Trip_Type'] = hotels_copy['Trip_Type'].str.lstrip("'")
hotels_copy['Trip_Type'] = hotels_copy['Trip_Type'].str.rstrip("'")


hotels_copy['Duration'] = hotels_copy['Duration'].str.replace(']', '')
hotels_copy['Duration'] = hotels_copy['Duration'].str.lstrip("' ")
hotels_copy['Duration'] = hotels_copy['Duration'].str.rstrip("'")

hotels_copy['Room'] = hotels_copy['Room'].str.lstrip("' ")
hotels_copy['Room'] = hotels_copy['Room'].str.rstrip("'")

hotels_copy['People'] = hotels_copy['People'].str.lstrip("' ")
hotels_copy['People'] = hotels_copy['People'].str.rstrip("'")


print("")
print('Printing the first four rows')
print(hotels_copy.head(4))

print("")
print('Printing the Trip_Type value_counts')
print(hotels_copy['Trip_Type'].value_counts())

print("")
print('Printing the People value_counts')
print(hotels_copy['People'].value_counts())

print("")
print('Printing the Room value_counts')
print(hotels_copy['Room'].value_counts())

print("")
print('Printing the Duration value_counts')
print(hotels_copy['Duration'].value_counts())

print("")
print('Printing the bottom three rows')
print(hotels_copy.tail(3))


Printing Columns
Index(['Hotel_Address', 'Additional_Number_of_Scoring', 'Average_Score',
       'Hotel_Name', 'Reviewer_Nationality', 'Negative_Review',
       'Review_Total_Negative_Word_Counts', 'Total_Number_of_Reviews',
       'Positive_Review', 'Review_Total_Positive_Word_Counts',
       'Total_Number_of_Reviews_Reviewer_Has_Given', 'Reviewer_Score', 'Tags',
       'days_since_review', 'lat', 'lng', 'Month', 'Year', 'Day', 'country',
       'Trip_Type', 'People', 'Room', 'Duration', 'Device', 'Other'],
      dtype='object')

Printing After Dropping the Columns
Index(['Hotel_Address', 'Additional_Number_of_Scoring', 'Average_Score',
       'Hotel_Name', 'Reviewer_Nationality', 'Negative_Review',
       'Review_Total_Negative_Word_Counts', 'Total_Number_of_Reviews',
       'Positive_Review', 'Review_Total_Positive_Word_Counts',
       'Total_Number_of_Reviews_Reviewer_Has_Given', 'Reviewer_Score',
       'days_since_review', 'lat', 'lng', 'Month', 'Year', 'Day', 'country',
       

C:\Users\hanna\AppData\Local\Temp/ipykernel_1880/4055156608.py:23: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  hotels_copy['Trip_Type'] = hotels_copy['Trip_Type'].str.replace('[', '')
C:\Users\hanna\AppData\Local\Temp/ipykernel_1880/4055156608.py:28: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  hotels_copy['Duration'] = hotels_copy['Duration'].str.replace(']', '')



Printing the first four rows
                                       Hotel_Address  \
0   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
1   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
2   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
3   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   

   Additional_Number_of_Scoring  Average_Score   Hotel_Name  \
0                           194            7.7  Hotel Arena   
1                           194            7.7  Hotel Arena   
2                           194            7.7  Hotel Arena   
3                           194            7.7  Hotel Arena   

  Reviewer_Nationality                                    Negative_Review  \
0              Russia    I am so angry that i made this post available...   
1             Ireland                                         No Negative   
2           Australia    Rooms are nice but for elderly a bit difficul...   
3      United Kingdom    My room was dirty and I was afraid to wa

We do see that the values are polluted. Values that may be missing in the first element, assumed to pick the second element as first, and third as second. We see that the Trip_Type values contain values of Trip_Type and values from 'People', 'People' contains value about people and the types of room. We see that extracting values using a string is better in this scenario whereas striping the characters to create it. 

In [120]:
#Mining some important information from tags
hotels['Leisure_trip'] = hotels['Tags'].map(lambda x: 1 if ' Leisure trip ' in x else 0)
hotels['Business_trip'] = hotels['Tags'].map(lambda x: 2 if ' Business trip ' in x else 0)
hotels['Trip_type'] = hotels['Leisure_trip'] + hotels['Business_trip']

In [121]:
hotels[5000:5005]

Hotel_Address  \
5000  1 8 Russell Square Camden London WC1B 5BE Unit...   
5001  1 8 Russell Square Camden London WC1B 5BE Unit...   
5002  1 8 Russell Square Camden London WC1B 5BE Unit...   
5003  1 8 Russell Square Camden London WC1B 5BE Unit...   
5004  1 8 Russell Square Camden London WC1B 5BE Unit...   

      Additional_Number_of_Scoring  Average_Score            Hotel_Name  \
5000                           724            8.0  The Principal London   
5001                           724            8.0  The Principal London   
5002                           724            8.0  The Principal London   
5003                           724            8.0  The Principal London   
5004                           724            8.0  The Principal London   

            Reviewer_Nationality  \
5000             United Kingdom    
5001             United Kingdom    
5002   United States of America    
5003             United Kingdom    
5004             United Kingdom    

                                        Negative_Review  \
5000                                        No Negative   
5001                                        No Negative   
5002    Tiny bathroom Breakfast ok but not very elegant   
5003   Room was too hot and quite compact Room overl...   
5004                                        No Negative   

      Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
5000                                  0                     3150   
5001                                  0                     3150   
5002                                  9                     3150   
5003                                 21                     3150   
5004                                  0                     3150   

                                        Positive_Review  \
5000   Thoughtfulness of staff fabulous room and ama...   
5001                                    Location staff    
5002             Wonderful location and grand old hotel   
5003                                 Location was great   
5004   Olde world lovely decor Convenient to tube an...   

      Review_Total_Positive_Word_Counts  ...  days_since_review        lat  \
5000                                 13  ...            644 day  51.522622   
5001                                  4  ...            645 day  51.522622   
5002                                  7  ...            646 day  51.522622   
5003                                  4  ...            646 day  51.522622   
5004                                 13  ...            646 day  51.522622   

          lng Month  Year  Day         country  Leisure_trip  Business_trip  \
5000 -0.12516    10  2015   29  United Kingdom             1              0   
5001 -0.12516    10  2015   28  United Kingdom             1              0   
5002 -0.12516    10  2015   27  United Kingdom             0              2   
5003 -0.12516    10  2015   27  United Kingdom             1              0   
5004 -0.12516    10  2015   27  United Kingdom             1              0   

     Trip_type  
5000         1  
5001         1  
5002         2  
5003         1  
5004         1  

[5 rows x 23 columns]

In [122]:
hotels['Trip_type'].value_counts()

1    417778
2     82939
0     15021
Name: Trip_type, dtype: int64

In [123]:
417778+82939

500717

In [124]:
417778/500717

0.8343595284362224

- Approach 1: Since we have 15k rows with 0, we can apply 83.4% of values to Leisure_trip and the rest of it Business_trip.
- Approach 2: We can delete these values, resulting in a loss of information. 

In [125]:
import random
hotels['Trip_0'] = hotels[hotels['Trip_type'] == 0]['Trip_type'].map(lambda x: 1 if random.random() > 0.2 else 2)
hotels['Trip_0'] = hotels['Trip_0'].fillna(0)
hotels['Trip_type'] = hotels['Trip_0'] + hotels['Business_trip'] + hotels['Leisure_trip']


We have distributed the values that were 0 to 1 and 2, however the rest were filled with NA. We can fillna(0) and then add Trip_0, Leisure_trip, and Business_trip

In [126]:
hotels['Trip_type'].value_counts()

1.0    429768
2.0     85970
Name: Trip_type, dtype: int64

Now we can do the same for the # of people, we see that we have many categories, however the most used ones are 
- Couple                                /        245754
- Solo traveler                         /        104060
- Group                                 /         63649
- Family with young children            /         58573
- Family with older children            /         25326

& this is from the second element of the list, the first element also contains information about the # of people.

In [127]:
245754 + 104060 + 63649 + 58573 + 25326

497362

These top 5 categories combine for almost 498k reviews, the rest we can fill in the blank. 

In [128]:
#Couple or Solo or Group or Family_with_older children or Family with younger Children
hotels['Solo'] = hotels['Tags'].map(lambda x: 1 if ' Solo traveler ' in x else 0)
hotels['Couple'] = hotels['Tags'].map(lambda x: 2 if ' Couple ' in x else 0)
hotels['Group'] = hotels['Tags'].map(lambda x: 3 if ' Group ' in x else 0)
hotels['Family_with_young_children'] = hotels['Tags'].map(lambda x: 4 if ' Family with young children ' in x else 0)
hotels['Family_with_older_children'] = hotels['Tags'].map(lambda x: 5 if ' Family with older children ' in x else 0)
hotels['Group_type'] = hotels['Solo'] + hotels['Couple'] + hotels['Group'] + hotels['Family_with_young_children'] + hotels['Family_with_older_children']

This process of extracting the information like this was found from a kaggle notebook. I found this way of extracting much easier to be dealt with than extracting by stripping the special characters, as in that way we found that the list had some elements missing and the index of elements moved back, hence making it harder to extract information from. 

The notebook can be found on: https://www.kaggle.com/code/mbkinaci/detailed-eda-and-xgboost-89-accuracy

In [129]:
hotels['Group_type'].value_counts()

2    252294
1    108545
3     65392
4     61015
5     26349
0      2143
Name: Group_type, dtype: int64

In [130]:
hotels[hotels['Group_type'] == 0].head(3)

Hotel_Address  \
20    s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
440  1 15 Templeton Place Earl s Court Kensington a...   
772  1 15 Templeton Place Earl s Court Kensington a...   

     Additional_Number_of_Scoring  Average_Score        Hotel_Name  \
20                            194            7.7       Hotel Arena   
440                           244            8.5  K K Hotel George   
772                           244            8.5  K K Hotel George   

    Reviewer_Nationality                                    Negative_Review  \
20       United Kingdom    Bed was on upper level with a narrow twist st...   
440               China    The wifi is only 500kb s which isn t fast by ...   
772               China                                             nothing   

     Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
20                                  40                     1403   
440                                 30                     1831   
772                                  2                     1831   

                                       Positive_Review  \
20    Friendly staff OostPark a few yards away Good...   
440   Generally it s a wonderful experience The hot...   
772   easy access to underground clean room with a ...   

     Review_Total_Positive_Word_Counts  ...  Leisure_trip  Business_trip  \
20                                  17  ...             1              0   
440                                137  ...             1              0   
772                                 13  ...             1              0   

    Trip_type Trip_0  Solo  Couple  Group  Family_with_young_children  \
20        1.0    0.0     0       0      0                           0   
440       1.0    0.0     0       0      0                           0   
772       1.0    0.0     0       0      0                           0   

     Family_with_older_children Group_type  
20                            0          0  
440                           0          0  
772                           0          0  

[3 rows x 30 columns]

Since Vacation are usually done with 2 people or more, and since Couple has the highest value in the # of people, we will assign the values that are 0 to 2. 

In [131]:
#I will assign 2 to 0's because it is the most.
hotels['Group_type'] = hotels['Group_type'].map(lambda x: 2 if x == 0 else x)
hotels['Group_type'].value_counts()

2    254437
1    108545
3     65392
4     61015
5     26349
Name: Group_type, dtype: int64

In [132]:
hotels.columns

Index(['Hotel_Address', 'Additional_Number_of_Scoring', 'Average_Score',
       'Hotel_Name', 'Reviewer_Nationality', 'Negative_Review',
       'Review_Total_Negative_Word_Counts', 'Total_Number_of_Reviews',
       'Positive_Review', 'Review_Total_Positive_Word_Counts',
       'Total_Number_of_Reviews_Reviewer_Has_Given', 'Reviewer_Score', 'Tags',
       'days_since_review', 'lat', 'lng', 'Month', 'Year', 'Day', 'country',
       'Leisure_trip', 'Business_trip', 'Trip_type', 'Trip_0', 'Solo',
       'Couple', 'Group', 'Family_with_young_children',
       'Family_with_older_children', 'Group_type'],
      dtype='object')

In [133]:
del hotels['Tags'], hotels['Leisure_trip'], hotels['Business_trip'], hotels['Trip_0'], hotels['Solo'], hotels['Couple'], hotels['Group'], hotels['Family_with_young_children'] , hotels['Family_with_older_children'] 

We have dropped the unnecessary columns because we got all the information in the Group_type, and Trip_type column.

In [134]:
hotels.head(3)

Hotel_Address  \
0   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
1   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   
2   s Gravesandestraat 55 Oost 1092 AA Amsterdam ...   

   Additional_Number_of_Scoring  Average_Score   Hotel_Name  \
0                           194            7.7  Hotel Arena   
1                           194            7.7  Hotel Arena   
2                           194            7.7  Hotel Arena   

  Reviewer_Nationality                                    Negative_Review  \
0              Russia    I am so angry that i made this post available...   
1             Ireland                                         No Negative   
2           Australia    Rooms are nice but for elderly a bit difficul...   

   Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
0                                397                     1403   
1                                  0                     1403   
2                                 42                     1403   

                                     Positive_Review  \
0   Only the park outside of the hotel was beauti...   
1   No real complaints the hotel was great great ...   
2   Location was good and staff were ok It is cut...   

   Review_Total_Positive_Word_Counts  ...  Reviewer_Score  days_since_review  \
0                                 11  ...             2.9             0 days   
1                                105  ...             7.5             0 days   
2                                 21  ...             7.1             3 days   

         lat       lng  Month  Year  Day      country Trip_type  Group_type  
0  52.360576  4.915968      8  2017    3  Netherlands       1.0           2  
1  52.360576  4.915968      8  2017    3  Netherlands       1.0           2  
2  52.360576  4.915968      7  2017   31  Netherlands       1.0           4  

[3 rows x 21 columns]

Some other columns we will not need as we move along forward, 
 - Additional Number of Scoring: This scoring is based on scores without the service, however since the values range from 0 - 2600, we can't really make sense of this range, hence we drop this column.
 - Review_Total_Negative_Word_Counts & Review_Total_Positive_Word_Counts: The number of words in a review doesn't help us visualize anything.

Let us see if we have any missing values and if so which column has the most. 

In [135]:
del hotels['Additional_Number_of_Scoring'], hotels['Review_Total_Negative_Word_Counts'], hotels['Review_Total_Positive_Word_Counts']

In [136]:
hotels.isnull().any().any()

True

In [137]:
missing = pd.concat([hotels.isnull().sum(), 100 * hotels.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='count', ascending=False)

count         %
lng                                          3268  0.633655
lat                                          3268  0.633655
Hotel_Address                                   0  0.000000
Average_Score                                   0  0.000000
Trip_type                                       0  0.000000
country                                         0  0.000000
Day                                             0  0.000000
Year                                            0  0.000000
Month                                           0  0.000000
days_since_review                               0  0.000000
Reviewer_Score                                  0  0.000000
Total_Number_of_Reviews_Reviewer_Has_Given      0  0.000000
Positive_Review                                 0  0.000000
Total_Number_of_Reviews                         0  0.000000
Negative_Review                                 0  0.000000
Reviewer_Nationality                            0  0.000000
Hotel_Name                                      0  0.000000
Group_type                                      0  0.000000

In [138]:
hotels.tail(3)

Hotel_Address  Average_Score  \
515735  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...            8.1   
515736  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...            8.1   
515737  Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...            8.1   

                   Hotel_Name Reviewer_Nationality  \
515735  Atlantis Hotel Vienna               Egypt    
515736  Atlantis Hotel Vienna              Mexico    
515737  Atlantis Hotel Vienna             Hungary    

                                          Negative_Review  \
515735   The ac was useless It was a hot week in vienn...   
515736                                        No Negative   
515737       I was in 3rd floor It didn t work Free Wife    

        Total_Number_of_Reviews  \
515735                     2823   
515736                     2823   
515737                     2823   

                                          Positive_Review  \
515735                                        No Positive   
515736   The rooms are enormous and really comfortable...   
515737                               staff was very kind    

        Total_Number_of_Reviews_Reviewer_Has_Given  Reviewer_Score  \
515735                                           3             2.5   
515736                                           3             8.8   
515737                                           1             8.3   

       days_since_review        lat        lng  Month  Year  Day  country  \
515735           715 day  48.203745  16.335677      8  2015   19  Austria   
515736           717 day  48.203745  16.335677      8  2015   17  Austria   
515737           725 day  48.203745  16.335677      8  2015    9  Austria   

        Trip_type  Group_type  
515735        1.0           5  
515736        1.0           3  
515737        1.0           4

In [139]:
hotels.to_csv('C:\\Users\\hanna\\OneDrive\\Desktop\\All Folders\\Data Science\\SpringBoard\\Capstone 3\\hotels_post_datawrangling.csv' , index=False)

A few interesting detail to account for.
- we have 1492 hotels, in 6 countries total, with United Kingdom has the most with 262301. 
- Almost 360k reviews were given by a single person (108k) & couples (252k) reviews.
- Almost 400k+ reviews came from a Leisure Trip, The rest were from business.
- We had to delete a few columns because of their repetitive nature.
- Our model building stage is based on the reviews column, however to extract the meaning of the type of trips, the scoring aspect of the hotels is something we will explore in Explatory Data Analysis notebook. 